Add several pages in section RTE (#1388)
TYPO3-Documentation/TYPO3CMS-Reference-CoreApi: Documentation/ApiOverview/Rte/InTheFrontend/Introduction.rst

.. include:: /Includes.rst.txt
.. index::
   pair: Rich text editor; Frontend
.. _rte-frontend-introduction:

==================================================
Including a Rich Text Editor (RTE) in the frontend
==================================================

When you add forms to a website you might want to offer
formatting options like bold, italic etc.. Rich Text Editors offer
extensive options that are configurable for individual needs.

This chapter outlines conceptual and technical information
about adding an RTE on frontend pages.

.. figure:: ../Images/RteBackend.png
   :alt: A RTE in the TYPO3 BE

>    The rtehtmlarea RTE activated in the TYPO3 backend

Reviewer: rtehtmlarea  is an old, outdated extension (https://extensions.typo3.org/extension/rtehtmlarea) let's use the rte_ckeditor here.

Author: That is the (wrong) caption of the image.  
I copied it from the backend chapter and it has to be replaced all together.  
So you're right but it's on the TODO-list anyway ;-)

.. tip::

>    This description about CKEditor includes examples and common challenges for

Reviewer: link to the system_extension manual?

   the frontend.
   
   You can use other editors with TYPO3 and some points, like
   handling of data on the server, are independent of the distinct editor
   in the frontend.

> .. note::

Reviewer: I would not include this note. Such a documentation would most likely belong in the extensions documentation. rte_ckeditor is currently the only one provided by the core

Author: You marked the text above the note, thats a bit confusing, but I refer to the note below that part.
For the frontend there is no need to use CKEditor, it's completely independent.  
In the manual of CKEditor I think this info is lost, it never belongs there in my opinion.  
Nevertheless, I could drop that note completely, people who feel the need for it or want to do something find a way to articulate it.

Author: unintentionally it was already gone before, no clue what happend

   If you feel able to provide a description about using other editors
   with TYPO3, please add to this documentation.

The following list describes features and corresponding implementation
effort ordered from simple to complex. 

The optional features
=====================

> #. Simple text formatting can be achieved using well-known buttons.

Reviewer: Why use an ordered list? I perceive no natural order here...

   other file-types could be displayed in the content, further file-types
   could be linked.

#. Editing data in the frontend requires that applying forms are pre-filled
   with existing data. This might require some considerations concerning
   multiple aspects.

#. Links might be choosen out of the existing pages of the website, those
   links can be added as internal instead of external links but require
   a visual and functional option to select from existing pages.
   
   *This option requires an AJAX-connection to interact with TYPO3.*

#. For special websites like intranets it might be desired additionally
   to not only allow the upload of media but also to choose media out of
   those that exist already in a public directory on the server.
   
   *This option requires an AJAX-connection to interact with TYPO3.*

.. important::

   User input in the frontend always represents a general security risk.

   **A Rich Text Editor might reduce this awareness** because it looks like a
   professional solution but in fact might increase vulnerability if features
   are enabled or included without proper handling on the server.

>    It's advised to allow only required input and to disallow any further

Reviewer: link to xss attack descriptions in security chapter? https://docs.typo3.org/m/typo3/reference-coreapi/master/en-us/Security/TypesOfThreats/Index.html#cross-site-scripting-xss

Author: yes!

.. include:: /Includes.rst.txt
.. index::
   pair: Rich text editor; Frontend
.. _rte-frontend-introduction:

==================================================
Including a Rich Text Editor (RTE) in the frontend
==================================================

> When forms are added to a website it is sometimes desired to offer
> formatting options like bold, italic etc.. Rich Text Editors offer
> here extensive options that are configurable for the individual needs.

Reviewer 2: ```suggestion
When you add forms to a website you might want to offer
formatting options like bold, italic etc.. Rich Text Editors offer
extensive options that are configurable for individual needs.
```

Author: Thanks @flicstar , I must admit my text is not so smooth, I try to avoid the "you" usually, so it changes everything by introducing it again.
I read now already all your other comments and you make things nicer to read, Thanks!!!
About the "you" in the moment I feel a bit undecided, but don't know if there exists some styleguide about it.

Reviewer 2: Good point, let me go check styling. I also need to check "a" or "an" RTE. 😁

> This chapter is aimed to give conceptual and technical information
> about the challenge to add a RTE on frontend-pages.

Reviewer 2: ```suggestion
This chapter outlines conceptual and technical information
about adding an RTE on frontend pages.
```

.. figure:: ../Images/RteBackend.png
   :alt: A RTE in the TYPO3 BE

   The rtehtmlarea RTE activated in the TYPO3 backend

.. tip::

>    The description about CKEditor includes challenges and examples for
>    the frontend.

Reviewer 2: ```suggestion
   This description about CKEditor includes examples and common challenges for
   the frontend.
```

>    Usage of other editors is nevertheless possible and some points like
>    handling of data on the server are independent of the distinct editor
>    in the frontend.

Reviewer 2: ```suggestion
   You can use other editors with TYPO3 and some points, like
   handling of data on the server, are independent of the distinct editor
   in the frontend.
```

.. note::

>    If you feel eligible to provide a description for other editors
>    than CKEditor, this could be included or linked in this documentation.

Reviewer 2: ```suggestion
   If you feel able to provide a description about using other editors
   with TYPO3, please add to this documentation.
```

> Efforts to include a RTE can differ very much depending on the offered
> features. The features and required efforts in the following list increase
> with every point.

Reviewer 2: ```suggestion
The following list describes features and corresponding implementation
effort ordered from simple to complex. 
```

The optional features
=====================

> #. Simple text formatting can be reached with help of a few simple buttons.
>    This solution is used to format text (bold, italic, underlined, ...),
>    create lists or tables, etc..
>    Character of these options is that they are predefined in the RTE and
>    and wrap selected content in html-tags, by default without any attributes
>    like `id`, `class` or more.

Reviewer 2: ```suggestion
#. Simple text formatting can be achieved using well-known buttons.
   This solution is used to format text (bold, italic, underlined, ...),
   create lists or tables, etc..
   These options are predefined in the RTE and
   and wrap selected content in html-tags, by default without any attributes
   like `id` or `class` for example.
```

   CKEditor offers a dropdown button for those block-styles.
   
   Editing the Source could allow the user optionally to add special
   HTML-elements or attributes like `id`, `class` or more.

#. It might be desired to allow users to upload files like images,
   PDF-documents or other multi-media-content. Images and perhaps some
   other file-types could be displayed in the content, further file-types
   could be linked.

#. Editing data in the frontend requires that applying forms are pre-filled
   with existing data. This might require some considerations concerning
   multiple aspects.

#. Links might be choosen out of the existing pages of the website, those
   links can be added as internal instead of external links but require
   a visual and functional option to select from existing pages.
   
   *This option requires an AJAX-connection to interact with TYPO3.*

#. For special websites like intranets it might be desired additionally
   to not only allow the upload of media but also to choose media out of
   those that exist already in a public directory on the server.
   
   *This option requires an AJAX-connection to interact with TYPO3.*

.. important::

>    User input in the frontend represents in general always a security risk.

Reviewer 2: ```suggestion
   User input in the frontend always represents a general security risk.
```

>    **A Rich Text Editor might reduce this awarness** as it looks like a
>    professional solution but in fact might increase the risk if features
>    are enabled or included without proper handling on the server.

Reviewer 2: ```suggestion
   **A Rich Text Editor might reduce this awareness** because it looks like a
   professional solution but in fact might increase vulnerability if features
   are enabled or included without proper handling on the server.
```

   It's advised to allow only required input and to disallow any further
   content and features.

Technical Overview
==================

**Files**: Any required files to include a form in the frontend require

> an extension, this can be a sitepackage but also a separated extension.

Reviewer 2: ```suggestion
an extension, this can be a sitepackage but also a separate extension.
```
